helm-v3/deisgn-proposal.md

```diff
@@ -0,0 +1,715 @@
+* A few in-cluster CRDs: Release, ReleaseVersion (Neither of which has an accompanying controller)
+
+In this model, Tiller is removed and there are no operators or controllers that run in-cluster.
+
+## Charts
+The chart architecture remains largely unchanged with the exception that the `Chart.yaml`
+file is now formatted as a Kubernetes resource.
+
+The following changes are proposed to the existing chart format:
+
+- Chart.yaml version 3
+- The overrides/ directory
+- The ext/ directory
+- Schemata for values files
+
+### Chart.yaml, Version 3
+
+In an effort to mirror the construction of Kubernetes manifests, the new Chart.yaml file looks like this:
+
+```yaml
+apiVersion: helm.sh/v3
+kind: Chart
+metadata:
+  name: myChart
+  labels:
+    heritage: helm
+    version: 0.1.0
+    appVersion: 3.6
+data:
+  description: Deploy a basic Alpine Linux pod
+  home: https://github.com/kubernetes/helm
+  sources:
+    - https://github.com/kubernetes/helm
+```
+
+This file contains only metadata that describes the application (not the application instance), so multiple instances of the same chart can be installed.
+
+The `heritage` label indicates which app created this. The CLI will use the heritage value of `helm`. Other components, such as an in-cluster controller, should specify their own value here.
+
+Commonly accessed properties are moved into labels, while less frequently used or more finely structured data is relegated to the `data:` section. The table below illustrates the transitions from [Helm v2](https://github.com/kubernetes/helm/blob/master/docs/charts.md#the-chartyaml-file) to Helm 3:
+
+
+| Helm 2 | Helm 3 | Notes |
+| -------- | -------- | -------- |
+| name     | metadata.name | |
+| version | metadata.labels.version | |
+| kubeVersion | data.kubeVersion | |
+| keywords | data.keywords | Represented as a YAML list frabgment|
+| description | data.description | |
+| home     | data.home | |
+| sources | data.sources | Represented as a YAML list fragment|
+| maintainers | data.maintainers | following the same object syntax as in Helm 2|
+| engine | REMOVED  | superseded by event system |
+| icon | data.icon | |
+| appVersion | metadata.labels.appVersion | Presumably, this is a field that should be queryable |
+| deprecated | data.deprecated | |
+| tillerVersion | REMOVED | There is no Tiller |
+
+For the sake of completeness, we will provide a CRD defining this type.
+
+*Note:* If the Application CRD proposal is workable during the course of early Helm 3 development, we will consider using that.
+
+### The overlays/ directory
+The `overlays/` directory contains templates that override templates in subcharts. This provides a mechanism for a high-level chart (like Wordpress) to override one or more individual templates from a subchart (like MySQL).
+
+> [name=Adnan Abdulhussein]
+> Is this a complete override, or some sort of merge? If it's a full override, then perhaps this feature should be called "overrides"?
```

Comment:
The `overlays` system has been removed since the initial commit because the functionality is provided by Lua (as is the merge functionality.)

I am trying to encourage people to view Lua-in-charts as a tool for modifying the resources (not so much as an extension mechanism). I wouldn't exactly call it a _replacement_ for the Go templates, but as the advanced option for writing something (a) more flexible, (b) componentized (Object oriented, if you prefer), and (c) much easier to read when the chart gets sophisticated. But at the same time, we can't simply remove Gotpl support. It remains the easiest way to get started, and there's also a long legacy that we want to support.

But as you noted, overlays look more like a band-aid than a solution. So given that the same thing can be accomplished with Lua, I just removed the entire overlay system.

Reply:
Yes, lua is more flexible. But do consider target audience. It is hard enough for me to ask end users to write yaml at times. If it can be simply done with yaml merging rather then having to switch to lua, it goes from needing a programmer to being more of a pure operator/scientist level. Does that make sense, or should I elaborate on the use case some more?

Reply:
`overlays` are a popular option. Kubectl folks are experimenting with it. Other configuration management systems use it. In addition to that, many ops folks are not developers and many developers are quite coupled to their programming language of choice rather than programming generalists. Do we want to make lua a requirement for overlays? if so, how do we make it easy? Just providing lua isn't going to make it easy or intuitive.